pkg/policy/api/cidr.go

```diff
@@ -83,12 +97,26 @@ func (s CIDRSlice) GetAsEndpointSelectors() EndpointSelectorSlice {
 	// we only have to add the EndpointSelector representing reserved:world
 	// once.
 	var hasWorldBeenAdded bool
+	var hasNodeBeenAdded bool
 	slice := EndpointSelectorSlice{}
 	for _, cidr := range s {
 		if cidr.MatchesAll() && !hasWorldBeenAdded {
 			hasWorldBeenAdded = true
 			slice = append(slice, ReservedEndpointSelectors[labels.IDNameWorld])
 		}
+		// If CIDR matches the local node IP, add the local host and remote node identity
+		// to the allowed list. This is a workaround to alleviate the CIDR issue
+		// https://github.com/cilium/cilium/issues/12277
+		// for kubernetes network policy.
+		// Local host and remote nodes are not distinguished to make the workaround simple.
+		// Note that this will change the behavior of cilium network policies as well if
+		// CIDR is configured and overlaps the node IP address.
+		if !hasNodeBeenAdded && cidr.MatchesLocalNode() {
+			hasNodeBeenAdded = true
+			slice = append(slice, ReservedEndpointSelectors[labels.IDNameHost])
+			slice = append(slice, ReservedEndpointSelectors[labels.IDNameRemoteNode])
```

Comment:
For the Host IP, this makes sense. The CIDR match would select the IP of the host and therefore allow (or in future, [deny](https://github.com/cilium/cilium/issues/7111)) the traffic based on the IP.

For the remote nodes, this doesn't make sense. injecting policies on each node that only select the *local node* IP for each node would open up traffic to *all nodes* in the cluster. That is a broader selection than what the policy states or what the average user would expect such a policy to allow. 

I feel like the right approach for the remote-nodes side here would be to only select remote-node identity if the CIDR covers the entire range of remote nodes in the cluster. I'm not sure what kind of information we get about the CIDR range used for nodes but one option could be `--native-routing-cidr`.

Reply:
Joe, thanks for the review! What do you mean by "injecting policies on each node"?  In the normal way of applying network policy, all nodes get the same copy. So if a CIDR contains only one node IP, even we distinguish local and remote, it will only work for one node. For the rest, they inevitably open more than they want because the system just doesn't have the resolution to distinguish individual remote node. 

It would be nice to know the node CIDR and only apply this rule if the cidr includes the whole node CIDR.  This will avoid accidentally misconfiguration causing non-consistent behavior across cluster nodes.  But I don't see a parameter carrying the information - native-routing-cidr looks more like cluster pod cidr from the definition here: https://github.com/cilium/cilium/blob/master/pkg/option/config.go#L710

This reminds me another issue - if node cidr is not included in native-routing-cidr, does pod to remote node traffic get masqueraded when masquerade is enabled? If yes, that can cause issue to host network policy. Is a list of native-routing-cidr supported?

In the long run, we probably should implement a way to apply a CIDR Policy regardless of who owns the IP. Allowing a CIDR would just literally allow all connections with IP in that CIDR. It provides a way to support whatever crazy use case customer may have.

Reply:
> What do you mean by "injecting policies on each node"?

I think you understood correctly my point here, by "policies" here I was considering the internal Cilium distilled policy representation that is determined based on the per-node context, not the k8s network policy level.

>  if node cidr is not included in native-routing-cidr, does pod to remote node traffic get masqueraded when masquerade is enabled?

Yes.

> In the long run, we probably should implement a way to apply a CIDR Policy regardless of who owns the IP. 

I recognize that it's rather opinionated from a policy perspective to encourage an alternative view than this given the history of computer networking and in this case, the assumption of `iptables` as the default networking configuration model, but this is intentional. The point of this approach is to push the long-term direction away from embedding location-based addressing in policies, towards identity-based polices where the policy is about semantically what you intend to access / who is allowed to to communicate to whom. With that abstraction in place, the concerns of routing/forwarding and addressing are separated from the concern of how the addresses map to the security identity of a workload, and this allows more efficient policy computation & packet handling. I'm a little rusty on the academic links to this but I recall it's related to ideas like Content-Addressable Networks and more recently Identifier-Locator Addressing. Long term this should both simplify the policy to allow it to be written with network forwarding intent in mind, while at the same time improving policy scalability & make it easier to manage the policies because they're not dependent on the addressing layouts of your clusters.

Reply:
Anyway, we had an out-of-band call where we found a potential path forward so with the 'remote-node' aspects of this PR removed I'm otherwise happy with the implementation.